File: src/engine/MeshPart.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Tao.OpenGl;

namespace SimpleEngine
{
    public class MeshPart
    {
        public List<MeshDef.Vertex[]> faces = new List<MeshDef.Vertex[]>();
        Material mat;
        Poseable poseable;

        public MeshPart(MediaLoader ml, MeshDef.MaterialDef m, MeshDef def, Poseable poseable)
        {
            this.poseable = poseable;

            mat = new Material(m.diffuse, ml.fetch<Texture>(m.texture));
            foreach (MeshDef.Tri tri in def.TrianglesFor(m))
            {
                faces.Add(def.lookup(tri));
            }
        }

        int id = -1;
        public int Id
        {
            get
            {
                if (id == -1) id = calculateId();
                return id;
            }
        }

        private int calculateId()
        {
            return 0;
        }
        internal void render(vec3 p, quat rot)
        {
            using(PushedMatrix dnum = new PushedMatrix() )
            {
                mat.apply();
                Gl.glTranslatef(p.x, p.y, p.z);
                AxisAngle aa = rot.AxisAngle;
                // apply rotation
                Gl.glRotatef(aa.angle.inDegrees, aa.axis.x, aa.axis.y, aa.axis.z);
                Gl.glBegin(Gl.GL_TRIANGLES);
                foreach (MeshDef.Vertex[] face in faces)
                {
                    for (int i = 0; i < 3; ++i)
                    {
                        MeshDef.Vertex v = face[i];
                        vec3 vert = (poseable.CurrentPose != null)
                            ? poseable.CurrentPose.transforms[v.bone] * v.vertex
                            : v.vertex;
                        Gl.glNormal3f(v.normal.x, v.normal.y, v.normal.z);
                        Gl.glTexCoord2f(v.uv.x, v.uv.y);
                        Gl.glVertex3f(vert.x, vert.y, vert.z);
                    }
                }
                Gl.glEnd();

            }
        }
    }
}

File: src/engine/meshpart.cc
#include "engine/meshpart.h"

#include <memory>
#include <vector>

#include "engine/animation.h"
#include "engine/axisangle.h"
#include "engine/medialoader.h"
#include "engine/opengl.h"
#include "engine/pushedmatrix.h"
#include "engine/texture.h"

namespace simple
{
    struct Poseable;

    MeshPart::MeshPart(MediaLoader* ml, const MaterialDefinition& m, MeshDef* def, Poseable* p)
        : poseable(p)
    {
        material = std::make_shared<Material>(m.diffuse, ml->GetOrLoad<TextureMedia>(m.texture));
        for (const auto& tri : def->GetTrianglesFor(m))
        {
            faces.emplace_back(def->GetVerticesForTriangle(tri));
        }
    }

    int MeshPart::GetId()
    {
        if (id == -1)
            id = CalculateId();
        return id;
    }

    int MeshPart::CalculateId() const
    {
        return 0;
    }

    void MeshPart::OnRender(const vec3& position, const quat& rotation) const
    {
        PushedMatrix pushed_matrix;

        material->Apply();
        glTranslatef(position.x, position.y, position.z);
        AxisAngle aa = rotation.GetAxisAngle();
        // apply rotation
        glRotatef(aa.angle.InDegrees(), aa.axis.x, aa.axis.y, aa.axis.z);
        glBegin(GL_TRIANGLES);
        for (const auto& face : faces)
        {
            for (auto i = 0; i < 3; ++i)
            {
                auto v = face[i];
                const auto vertex = (poseable->GetPose() != nullptr)
                                      ? poseable->GetPose()->transforms[v.bone] * v.position
                                      : v.position;
                glNormal3f(v.normal.x, v.normal.y, v.normal.z);
                glTexCoord2f(v.texture_coordinate.x, v.texture_coordinate.y);
                glVertex3f(vertex.x, vertex.y, vertex.z);
            }
        }
        glEnd();
    }
}
